lib/autoupdate/tools/updater.go

```diff
@@ -0,0 +1,390 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package tools
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"os"
+	"os/exec"
+	"os/signal"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"syscall"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/client/webclient"
+	"github.com/gravitational/teleport/api/constants"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/teleport/lib/utils/packaging"
+)
+
+const (
+	// teleportToolsVersionEnv is environment name for requesting specific version for update.
+	teleportToolsVersionEnv = "TELEPORT_TOOLS_VERSION"
+	// baseURL is CDN URL for downloading official Teleport packages.
+	baseURL = "https://cdn.teleport.dev"
+	// reservedFreeDisk is the predefined amount of free disk space (in bytes) required
+	// to remain available after downloading archives.
+	reservedFreeDisk = 10 * 1024 * 1024 // 10 Mb
+	// lockFileName is file used for locking update process in parallel.
+	lockFileName = ".lock"
+	// updatePackageSuffix is directory suffix used for package extraction in tools directory.
+	updatePackageSuffix = "-update-pkg"
+)
+
+var (
+	// // pattern is template for response on version command for client tools {tsh, tctl}.
+	pattern = regexp.MustCompile(`(?m)Teleport v(.*) git`)
+)
+
+// Option applies an option value for the Updater.
+type Option func(u *Updater)
+
+// WithBaseURL defines custom base url for the updater.
+func WithBaseURL(baseUrl string) Option {
+	return func(u *Updater) {
+		u.baseUrl = baseUrl
+	}
+}
+
+// WithClient defines custom http client for the Updater.
+func WithClient(client *http.Client) Option {
+	return func(u *Updater) {
+		u.client = client
+	}
+}
+
+// Updater is updater implementation for the client tools auto updates.
+type Updater struct {
+	toolsDir     string
+	localVersion string
+	tools        []string
+
+	baseUrl string
+	client  *http.Client
+}
+
+// NewUpdater initiate updater for the client tools auto updates.
```

Comment:
Can you document what each of the arguments takes? E.g., what format should be used for `tools`?